File: euler/python/1.py
problem = '''Find the sum of all the multiples of 3 or 5 below 1000. \n'''

# With nested func
def w_nested(limit):
  def criteria(n):
    return n % 3 == 0 or n % 5 == 0
  return sum(list(filter(criteria, range(1, limit))))

# With lambda
def w_lambda(limit):
  return sum(list(filter(lambda n: n % 3 == 0 or n % 5 == 0, range(1, limit))))

print(problem)
print(w_nested(1000))
print(w_lambda(1000))

File: euler/rust/1.rs

use std::iter::AdditiveIterator;

fn main() {
  let problem = "1: Multiples of 3 and 5";
  println!("Euler problem {} \n", problem);
  // Begin problem
  println!("Procedural: {}", procedural());
  println!("Functional: {}", functional());
}

fn functional() -> int {
  let it = range(0i, 1000);
  it.filter(|n| n % 3 == 0 || n % 5 == 0).sum()
}

fn procedural() -> int {
  let mut sum = 0i;
  for i in range(0i, 1000) {
    if i % 3 == 0 || i % 5 == 0 {
      sum = sum + i;
    }
  }
  sum
}

#[test]
fn test_problem() {
  let p = procedural();
  let f = functional();
  assert!(p == 233168);
  assert!(f == 233168);
}

